sdk/src/azure/core/az_mqtt5_rpc_server_hfsm.c

```diff
@@ -0,0 +1,598 @@
+
+AZ_INLINE az_result _handle_request(az_mqtt5_rpc_server* me, az_mqtt5_recv_data* data);
+AZ_NODISCARD az_result _az_rpc_server_policy_init(
+    _az_hfsm* hfsm,
+    _az_event_client* event_client,
+    az_mqtt5_connection* connection);
+
+static az_event_policy_handler _get_parent(az_event_policy_handler child_state)
+{
+  az_event_policy_handler parent_state;
+
+  if (child_state == root)
+  {
+    parent_state = NULL;
+  }
+  else if (child_state == subscribing || child_state == waiting || child_state == faulted)
+  {
+    parent_state = root;
+  }
+  else
+  {
+    // Unknown state.
+    az_platform_critical_error();
+    parent_state = NULL;
+  }
+
+  return parent_state;
+}
+
+static az_result root(az_event_policy* me, az_event event)
+{
+  az_result ret = AZ_OK;
+  (void)me;
+
+  if (_az_LOG_SHOULD_WRITE(event.type))
+  {
+    _az_LOG_WRITE(event.type, AZ_SPAN_FROM_STR("az_rpc_server"));
+  }
+
+  switch (event.type)
+  {
+    case AZ_HFSM_EVENT_ENTRY:
+      break;
+
+    case AZ_HFSM_EVENT_ERROR:
+    {
+      if (az_result_failed(az_event_policy_send_inbound_event(me, event)))
+      {
+        az_platform_critical_error();
+      }
+      break;
+    }
+
+    case AZ_HFSM_EVENT_EXIT:
+    {
+      if (_az_LOG_SHOULD_WRITE(AZ_HFSM_EVENT_EXIT))
+      {
+        _az_LOG_WRITE(AZ_HFSM_EVENT_EXIT, AZ_SPAN_FROM_STR("az_mqtt5_rpc_server: PANIC!"));
+      }
+
+      az_platform_critical_error();
+      break;
+    }
+
+    case AZ_MQTT5_EVENT_PUBACK_RSP:
+    case AZ_EVENT_MQTT5_CONNECTION_OPEN_REQ:
+    case AZ_MQTT5_EVENT_CONNECT_RSP:
+    case AZ_EVENT_MQTT5_CONNECTION_CLOSE_REQ:
+    case AZ_MQTT5_EVENT_DISCONNECT_RSP:
+      break;
+
+    default:
+      // TODO
+      ret = AZ_HFSM_RETURN_HANDLE_BY_SUPERSTATE;
+      break;
+  }
+
+  return ret;
+}
+
+/**
+ * @brief start subscription timer
+ */
+AZ_INLINE az_result _rpc_start_timer(az_mqtt5_rpc_server* me)
+{
+  _az_event_pipeline* pipeline = &me->_internal.connection->_internal.event_pipeline;
+  _az_event_pipeline_timer* timer = &me->_internal.rpc_server_memory._internal.rpc_server_timer;
+
+  _az_RETURN_IF_FAILED(_az_event_pipeline_timer_create(pipeline, timer));
+
+  int32_t delay_milliseconds
+      = (int32_t)me->_internal.rpc_server_memory._internal.subscribe_timeout_in_seconds * 1000;
+  if (delay_milliseconds <= 0)
+  {
+    delay_milliseconds = AZ_MQTT5_RPC_SERVER_DEFAULT_TIMEOUT_SECONDS * 1000;
+  }
+
+  _az_RETURN_IF_FAILED(az_platform_timer_start(&timer->platform_timer, delay_milliseconds));
+
+  return AZ_OK;
+}
+
+/**
+ * @brief stop subscription timer
+ */
+AZ_INLINE az_result _rpc_stop_timer(az_mqtt5_rpc_server* me)
+{
+  _az_event_pipeline_timer* timer = &me->_internal.rpc_server_memory._internal.rpc_server_timer;
+  return az_platform_timer_destroy(&timer->platform_timer);
+}
+
+/**
+ * @brief helper function to check if an az_span topic matches an az_span subscription
+ */
+AZ_NODISCARD AZ_INLINE bool az_span_topic_matches_sub(az_span sub, az_span topic)
```

Comment:
No, because if the subscription has wildcards it won't match the incoming topic exactly :( I will be creating our own version of this function though instead of using the one from mosquitto